### PR TITLE
fix(security): allow vscode/cursor/zed deep-links in openExternal

### LINKED ISSUE #14433: [Bug]:  "Open in VS Code" button blocked by Electron security for vscode:// protocol
### Issue Checklist

- [x] I understand that issues are for feedback and problem solving, not for complaining in the comment section, and will provide as much information as possible to help solve the problem.
- [x] My issue is not listed in the [FAQ](https://github.com/CherryHQ/cherry-studio/issues/3881).
- [x] I've looked at **pinned issues** and searched for existing [Open Issues](https://github.com/CherryHQ/cherry-studio/issues), [Closed Issues](https://github.com/CherryHQ/cherry-studio/issues?q=is%3Aissue%20state%3Aclosed), and [Discussions](https://github.com/CherryHQ/cherry-studio/discussions), no similar issue or discussion was found.
- [x] I've filled in short, clear headings so that developers can quickly identify a rough idea of what to expect when flipping through the list of issues. And not "a suggestion", "stuck", etc.
- [x] I've confirmed that I am using the latest version of Cherry Studio.

### Platform

Windows

### Version

1.9.2

### Bug Description

The "Open in VS Code" button in Cherry Studio interface does not work. When clicked, nothing happens. The logs show Electron security blocking the vscode:// protocol as an "untrusted URL scheme".

Actual Behavior
Button click produces no visible response. Cherry Studio logs show security blocks.

### Steps To Reproduce

1. Open Cherry Studio
2. Navigate to any agent
3. Click the "Open in VS Code" button
4. Observe: nothing happens
5. Check logs for Blocked shell.openExternal for untrusted URL scheme: vscode://

### Expected Behavior

Clicking the "Open in VS Code" button should open the current project in VS Code.

### Relevant Log Output

```shell
{
  "level": "warn",
  "message": "Blocked shell.openExternal for untrusted URL scheme: vscode://file/C%3A/Users/user/AppData/Roaming/CherryStudio/Data/Agents/t-default?windowId=_blank",
  "module": "WindowService",
  "process": "main",
  "timestamp": "2026-04-21 01:17:47"
}
```

### Additional Context

Important Diagnostic Information

1. Link works in Chrome: The exact URL Cherry Studio generates (vscode://file/C:/Users/user/AppData/Roaming/CherryStudio/Data/Agents/t-default?windowId=_blank) successfully opens VS Code when pasted directly into Chrome's address bar

2. Problem is Cherry Studio/Electron-specific: The link works system-wide, but Cherry Studio's Electron security blocks it internally

3. Potential symlink issue: Cherry Studio data is stored via symbolic links; the generated URL uses the original path rather than the resolved symlink path

Proposed Solution
Add vscode:// to the list of trusted URL schemes in Cherry Studio's Electron configuration, or implement a configurable allowlist for external protocols.

### PR DESCRIPTION
**Pull request by beyondkmp**
### What this PR does

fix #14433

Before this PR:

- The external-app config in `packages/shared/externalApp/config/index.ts` exposes **Visual Studio Code / Cursor / Zed** as code-editor targets, and the renderer has "Open in <editor>" buttons on the Agent chat navbar (`OpenExternalAppButton.tsx`) and inline file paths in agent messages (`ClickableFilePath.tsx`) that call `window.open(buildEditorUrl(app, path))`.
- These eventually hit `mainWindow.webContents.setWindowOpenHandler` in `src/main/services/WindowService.ts`, which checks the URL against `isSafeExternalUrl()` before calling `shell.openExternal`.
- `isSafeExternalUrl()`'s allowlist only contained `http:`, `https:`, `mailto:`, `obsidian:`, so `vscode://…`, `cursor://…`, `zed://…` all fell through to the `else` branch and were silently denied with a `Blocked shell.openExternal for untrusted URL scheme: vscode://…` warning. Clicking "Open in VS Code" did nothing.

After this PR:

- `ALLOWED_EXTERNAL_PROTOCOLS` also includes `vscode:`, `vscode-insiders:`, `cursor:`, `zed:`, so the deep-links produced by `buildEditorUrl()` pass the safety check and actually open the file in the chosen editor.
- `security.test.ts` gains an `allows code-editor deep-link protocols` case covering all four schemes (including the exact URL shape reported in the logs: `vscode://file/C%3A/…?windowId=_blank`).

Fixes #

### Why we need it and why it was done in this way

Users already see and click the editor buttons, so the current behavior (silent failure + warning log) is a functional regression, not a security-hardening feature. We just forgot to sync the main-process allowlist when those editor buttons were introduced.

The following tradeoffs were made:

- Kept the fix minimal — only the allowlist and its test are touched — per the `main`-branch code-freeze rule that hotfix PRs must not include refactoring.
- Added `vscode-insiders:` proactively so VS Code Insiders users won't hit the same dead-button bug even though the current config doesn't list Insiders as a separate editor.

The following alternatives were considered:

- Deriving the allowlist dynamically from `EXTERNAL_APPS` in `packages/shared/externalApp/config` so new editors are auto-covered. Rejected for now because it's a refactor and `main` is frozen; suitable for the `v2` branch instead.

Links to places where the discussion took place: N/A

### Breaking changes

None.

### Special notes for your reviewer

- Security-sensitive surface. The added schemes are all documented editor deep-links (`vscode://`, `vscode-insiders://`, `cursor://`, `zed://`) that the app already offers UI for. No `file:`, `javascript:`, `ms-msdt:`, etc. are permitted — the existing negative tests still pass.
- The same `isSafeExternalUrl()` is also used by `will-navigate` and `WebviewService` / `ipc.ts` `openExternal`, so the newly-allowed schemes are consistent across entry points (and that's the behavior we want — opening a file in VS Code from a message link should work everywhere).

### Checklist

- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. Check this only when the PR introduces or changes a user-facing feature or behavior.
- [x] Self-review: I have reviewed my own code (e.g., via [`/gh-pr-review`](/.claude/skills/gh-pr-review/SKILL.md), `gh pr diff`, or GitHub UI) before requesting review from others

### Release note

```release-note
Fix "Open in VS Code / Cursor / Zed" buttons on the Agent chat navbar and inline file-path menus being silently blocked by the main-process URL allowlist. Clicking an editor button now actually opens the file in the selected editor.
```

### CODE REVIEW

**Review by eeee0717 — APPROVED** [review]

**Comment by beyondkmp** [comment]
Hardened in f652de905 per review feedback.

`isSafeExternalUrl()` now also requires `host === 'file'` for the four editor schemes, so only the exact shape `buildEditorUrl()` emits — `<scheme>://file/...` — passes. Non-file authorities like `vscode://command/<id>` (runs registered editor commands) and `vscode://<publisher>.<ext>/...` (extension URL handlers) are rejected, which removes the broader attack surface from clickable links in chat/agent output.

Added negative tests covering:
- `vscode://command/workbench.action.terminal.sendSequence?text=rm`
- `vscode://ms-python.python/do-something`
- `cursor://settings`
- `zed://extension/evil`
- opaque form `vscode:command/foo`

**Comment by beyondkmp** [comment]
Regression fix pushed in 8a60ca4d0.

**Root cause**: `buildEditorUrl()` emits Zed URLs as ``${app.protocol}file${encodedPath}`` with no `/` separator. On Unix the encoded path starts with `/`, so the authority parses as `file` and the previous `host === 'file'` check was fine. On Windows the encoded path starts with a drive letter (e.g. `C%3A/Users/...`), producing `zed://fileC%3A/Users/...` — authority parses as `fileC%3A`, so the hardened check wrongly blocked Zed on Windows.

**Fix**: for `zed:` specifically, match the raw URL against the two concrete shapes `buildEditorUrl()` can emit:
- `zed://file/...` — Unix absolute path
- `zed://file<letter>%3A/...` — Windows drive letter, URL-encoded

Implemented as a small regex ``/^zed:\/\/file(\/|[A-Za-z]%3[Aa]\/)/i``. Also added a blanket rejection of any URL carrying userinfo to close a smuggling vector like `zed://file@evil.com/path` where `file` ends up as the username and the real host is attacker-controlled.

`vscode:` / `vscode-insiders:` / `cursor:` keep the stricter `host === 'file'` check since they always emit `<scheme>://file/<path>`.

Added tests:
- positive: `zed://fileC%3A/Users/foo/bar.ts`, `zed://filed%3a/data/foo.ts`
- negative: `zed://filename/path`, `zed://files.evil.com/cmd`, `zed://file@evil.com/path`, `vscode://file:pw@evil.com/path`

12/12 security tests pass.

**Review by DeJeune — APPROVED** [review]

**Comment by kangfenmao** [comment]
> [\!NOTE]
> This issue/comment/review was translated by Claude.

Could you also add this rule to window.api.shell.openExternal to avoid executing exe files in the operating system?

---
<details>
<summary>Original Content</summary>

能否再把这个规则加入到 window.api.shell.openExternal 里面，这样能避免去执行操作系统内的 exe 文件

</details>

File: src/main/services/security.ts [hunk 1]
```diff
@@ -2,22 +2,73 @@
...
+/**
+ * Zed's deep-link format is `zed://file<path>` (no slash separator before
+ * the path — Zed strips the `zed://file` prefix and treats the rest as a
+ * filesystem path). That means on Unix the URL is `zed://file/abs/path`
+ * (host parses as `file`), but on Windows it is `zed://fileC%3A/abs/path`
+ * (host parses as `fileC%3A`), so a plain `host === 'file'` check is
+ * insufficient. Match the two exact shapes buildEditorUrl() can emit: a
+ * slash, or a single-letter encoded drive followed by a slash.
+ */
+const ZED_FILE_URL_RE = /^zed:\/\/file(\/|[A-Za-z]%3[Aa]\/)/i
```
Comment (GeorgeDong32) [inline src/main/services/security.ts:34]: > [\!NOTE]
> This comment was translated by Claude.

**Nit: Redundancy between regex `i` flag and `%3[Aa]`**

Both `%3A` and `%3a` are already covered by `[Aa]`, so the outer `i` flag has no additional effect here (URL percent-encoding is not affected by case-insensitive matching). This isn't a problem, just noting it—keeping the `i` flag is harmless and acceptable as defensive coding.

---
<details>
<summary>Original Content</summary>

**Nit: 正则的 `i` flag 与 `%3[Aa]` 存在冗余**

`%3A` 和 `%3a` 已被 `[Aa]` 覆盖，外层 `i` flag 在这里没有额外效果（URL percent-encoding 不受 case-insensitive 影响）。不算问题，只是提一下——保留 `i` 也不伤，作为防御性编码可以接受。

</details>

File: src/main/services/security.ts [hunk 2]
```diff
@@ -2,22 +2,73 @@
...
+  if (parsed.username !== '' || parsed.password !== '') {
+    return false
+  }
+  if (parsed.protocol === 'zed:') {
+    return ZED_FILE_URL_RE.test(rawUrl)
+  }
+  // vscode / vscode-insiders / cursor all produce <scheme>://file/<path>,
+  // where the URL authority is exactly "file".
+  return parsed.host === 'file'
+}
```
Comment (GeorgeDong32) [inline src/main/services/security.ts:74]: > [\!NOTE]
> This comment was translated by Claude.

**Design Highlight: Excellent Two-Stage Validation Architecture**

`isFileOpenEditorUrl()` implements three layers of protection for editor schemes:
1. Userinfo rejection (prevents host injection)
2. Zed-specific regex validation (handles separator-less format + Windows drive letters)
3. Strict `host === 'file'` matching for VS Code/Cursor

This ensures that even if the allowlist permits the scheme, attackers cannot exploit dangerous authorities like `vscode://command/...`. The security design is very rigorous.

---
<details>
<summary>Original Content</summary>

**设计亮点：二次校验架构很棒**

`isFileOpenEditorUrl()` 对编辑器 scheme 做了三层防护：
1. Userinfo 拒绝（防 host 注入）
2. Zed 特殊正则校验（处理无分隔符格式 + Windows 驱动器号）
3. VS Code/Cursor 的 `host === 'file'` 严格匹配

这确保了即使白名单放行了 scheme，攻击者也无法利用 `vscode://command/...` 等危险 authority。安全设计非常严谨。

</details>

File: src/main/services/__tests__/security.test.ts [hunk 3]
```diff
@@ -21,6 +21,40 @@ describe('isSafeExternalUrl', () => {
...
+  })
+
+  it('rejects Zed deep-links that do not match the file-open shape', () => {
+    // host starts with "file" but is not file or file<drive>
+    expect(isSafeExternalUrl('zed://filename/path')).toBe(false)
+    expect(isSafeExternalUrl('zed://files.evil.com/cmd')).toBe(false)
+    // userinfo smuggling: "file" in userinfo, real host is attacker-controlled
+    expect(isSafeExternalUrl('zed://file@evil.com/path')).toBe(false)
+    expect(isSafeExternalUrl('vscode://file:pw@evil.com/path')).toBe(false)
+  })
```
Comment (GeorgeDong32) [inline src/main/services/__tests__/security.test.ts:56]: > [\!NOTE]
> This comment was translated by Claude.

**Comprehensive Test Coverage**

The edge case testing for Zed deep-links here is spot-on:
- `filename/path` — prefix spoofing
- `files.evil.com/cmd` — subdomain bypass
- `file@evil.com/path` — userinfo smuggling
- `file:pw@evil.com/path` — password injection

Especially the userinfo smuggling case, which is a typical URL parser differential attack vector. Kudos.

---
<details>
<summary>Original Content</summary>

**测试覆盖全面**

这里对 Zed deep-link 的边界 case 测得很到位：
- `filename/path` — 前缀欺骗
- `files.evil.com/cmd` — 子域名绕过
- `file@evil.com/path` — userinfo smuggling
- `file:pw@evil.com/path` — password 注入

尤其是 userinfo smuggling 这个 case，是很典型的 URL parser 差异攻击向量。赞。

</details>

**Review by GeorgeDong32 — APPROVED** [review]
## Review Summary: **Approve** ✅

High-quality security hotfix. Minimal scope, rigorous validation, thorough tests.

### Security ✅
- Protocol allowlist extension is well-scoped (editor deep-links only)
- `isFileOpenEditorUrl()` adds a second layer of defense: rejects `vscode://command/...`, extension handlers, and userinfo smuggling
- No dangerous schemes (`file:`, `javascript:`, `ms-msdt:`) introduced

### Code Quality ✅
- Minimal changes — only `security.ts` + tests, compliant with main branch code freeze
- Clear, well-documented code with helpful inline comments explaining design rationale
- Correct handling of Zed's quirky `zed://file<path>` format including Windows drive letters

### Testing ✅
- 7 → 12 tests (+5 new cases)
- Positive + negative coverage for all 4 schemes
- Edge cases covered: userinfo injection, host prefix spoofing, subdomain bypass, missing authority

### Non-blocking notes
- `obsidian:` remains unrestricted (pre-existing behavior, out of scope)
- `vscode-insiders:` added proactively despite not being in `EXTERNAL_APPS` config — good defensive call

**Comment by beyondkmp** [comment]
Worth flagging that the `.exe` concern is actually already covered on this path. The renderer's `window.api.shell.openExternal` doesn't go through IPC / `isSafeExternalUrl` — it's implemented inline in `src/preload/index.ts:434-448` with its own allowlist `['http:', 'https:', 'mailto:', 'obsidian:']` and calls `shell.openExternal` directly from preload:

```ts
shell: {
  openExternal: (url, options) => {
    const ALLOWED_PROTOCOLS = ['http:', 'https:', 'mailto:', 'obsidian:']
    try {
      const parsed = new URL(url)
      if (!ALLOWED_PROTOCOLS.includes(parsed.protocol)) {
        return Promise.reject(new Error(`Blocked openExternal for untrusted URL scheme: ${parsed.protocol}`))
      }
    } catch {
      return Promise.reject(new Error('Blocked openExternal for invalid URL'))
    }
    return shell.openExternal(url, options)
  }
}
```

So `file://…calc.exe`, `ms-msdt:`, `javascript:`, etc. are already blocked through this API today — the renderer cannot launch arbitrary `.exe`.

That said, the two allowlists have drifted:

- `src/main/services/security.ts` (this PR): http/https/mailto/obsidian + vscode/vscode-insiders/cursor/zed, with host & userinfo hardening
- `src/preload/index.ts`: http/https/mailto/obsidian only, no host/userinfo check

No renderer currently calls `window.api.shell.openExternal('vscode://...')` (editor deep-links go through `window.open()`, which hits the main-process `setWindowOpenHandler`), so the drift is latent. There is one related side effect unrelated to this PR: `src/renderer/src/pages/settings/SelectionAssistantSettings/components/MacProcessTrustHintModal.tsx:17` opens `x-apple.systempreferences:com.apple.preference.security?Privacy_Accessibility`, which is silently rejected by this same preload allowlist.

Unifying the two is a worthwhile **consistency** improvement, but it isn't plugging an open hole. Two ways:

1. **Inline sync** — replicate `isSafeExternalUrl`'s logic (host check, userinfo rejection, Zed regex) into `src/preload/index.ts`. Minimal; re-introduces drift risk next time the allowlist changes.
2. **Share the helper** — move `security.ts` into `packages/shared/` and import from both preload and main (plus `ipc.ts`, `WindowService`, `WebviewService`). Single source of truth, but it's a file move / cross-package refactor.

Per the CLAUDE.md main-branch rule that hotfix PRs must be minimal and non-refactoring, I'd keep this PR focused on the editor deep-link bug and do the unification as a follow-up. Happy to send either:

- a follow-up `hotfix/*` PR with option 1 (pure additive, no file moves), or
- a v2 PR with option 2, bundling the `x-apple.systempreferences:` oversight above in one shot.

Which way would you prefer?
